lib/BitstampStream.js

```diff
@@ -0,0 +1,95 @@
+"use strict";
+
+const EventEmitter = require("events");
+const ReconnectingWebSocket = require("reconnecting-websocket")
+const WebSocket = require("ws");
+
+class BitstampStream extends EventEmitter {
+    EVENT_UNSUBSCRIPTION_SUCCEEDED = "bts:unsubscription_succeeded";
+    EVENT_SUBSCRIPTION_SUCCEEDED = "bts:subscription_succeeded";
+    EVENT_REQUEST_RECONNECT = "bts:request_reconnect";
+
+    CHANNEL_LIVE_TRADES = "live_trades";
+    CHANNEL_LIVE_ORDERS = "live_orders";
+    CHANNEL_ORDER_BOOK = "order_book";
+    CHANNEL_DETAIL_ORDER_BOOK = "detail_order_book";
+    CHANNEL_DIFF_ORDER_BOOK = "diff_order_book";
+
+    constructor(url="wss://ws.bitstamp.net"){
+        super();
+        this.url = url;
+
+        this.connect();
+        this.ws.addEventListener("open", () => {
+            super.emit("connected");
+        });
+        this.ws.addEventListener("close", () => {
+            super.emit("disconnected");
+        });
+        this.ws.addEventListener("message", this._onMessageHandler.bind(this));
+    }
+
+    subscribe(channelBase, currencyPair){
+        const channel = `${channelBase}_${currencyPair}`;
+        this._send("bts:subscribe", { channel });
+        return channel;
+    }
+
+    unsubscribe(channelBase, currencyPair){
+        const channel = `${channelBase}_${currencyPair}`;
+        this._send("bts:unsubscribe", { channel });
+    }
+
+    unsubscribeAll(){
+        for (const channel of this.subscriptions) {
+            this._send("bts:unsubscribe", { channel });
+        }
+    }
+
+    connect(){
+        const options = {
+            WebSocket, // custom WebSocket constructor
+            connectionTimeout: 1000,
+            maxRetries: 10,
+            debug: true,
+        };
+        this.ws = new ReconnectingWebSocket(this.url, [], options);
+        this.subscriptions = new Set()
+    }
+
+    reconnect(){
+        this.close();
+        this.connect();
+    }
+
+    _send(event, data){
+        this.ws.send(JSON.stringify({ event, data }));
+    }
+
+    _onMessageHandler(messageEvent){
+        if (messageEvent.data) {
+            let { data, event, channel } = JSON.parse(messageEvent.data);
```

Comment:
potentially want to wrap this stuff in try catch in case something goes south and we can emit an error event

Reply:
Of course, you're right. I added a try catch and emit an `error` event.

Also cleaned up examples. Sorry about the messy PR. Seems I missed some details in examples as I did a minor refactor before pushing with `subscribe` method signature.

Reply:
Oh, and forgot to link issue. This addresses #29